bundles/org.openhab.binding.tibber/src/main/java/org/openhab/binding/tibber/internal/handler/TibberHandler.java

```diff
@@ -0,0 +1,471 @@
+            this.pollingJob = null;
+        }
+        if (isConnected()) {
+            close();
+        }
+        super.dispose();
+    }
+
+    public void open() throws Exception {
+        if (isConnected()) {
+            logger.debug("Open: connection is already open");
+        } else {
+            logger.info("Connecting to: {}", SUBSCRIPTION_URL);
+            TibberConfiguration tibberConfig = getConfigAs(TibberConfiguration.class);
+
+            sslContextFactory.setTrustAll(true);
+            sslContextFactory.setEndpointIdentificationAlgorithm(null);
+
+            WebSocketClient newClient = new WebSocketClient(sslContextFactory);
+            newClient.setMaxIdleTimeout(360 * 1000);
+
+            socket = new TibberWebSocketListener();
+
+            ClientUpgradeRequest newRequest = new ClientUpgradeRequest();
+
+            String token = new StringBuilder("Bearer ").append(tibberConfig.getToken()).toString();
+            newRequest.setHeader("Authorization", token);
+            newRequest.setSubProtocols("graphql-subscriptions");
+
+            newClient.start();
+            newClient.connect(socket, new URI(SUBSCRIPTION_URL), newRequest);
+        }
+    }
+
+    @SuppressWarnings("null")
+    public void close() {
+        Session session = this.session;
+        if (session != null) {
+            String disconnect = "{\"type\":\"connection_terminate\",\"payload\":null}";
+            try {
+                socket.sendMessage(disconnect);
+            } catch (IOException e) {
+                logger.warn("Closing websocket stream");
+            }
+            session.close();
+            this.session = null;
+        }
+    }
+
+    public boolean isConnected() {
+        Session session = this.session;
+        if (session == null || !session.isOpen()) {
+            return false;
+        }
+        return connected;
+    }
+
+    @WebSocket
+    @NonNullByDefault
+    public class TibberWebSocketListener {
+
+        @OnWebSocketConnect
+        public void onConnect(Session wssession) {
+            TibberConfiguration tibberConfig = getConfigAs(TibberConfiguration.class);
+            session = wssession;
+            connected = true;
+            logger.info("Connected to Server");
+
+            String connection = "{\"type\":\"connection_init\", \"payload\":\"token=" + tibberConfig.getToken() + "\"}";
+            try {
+                sendMessage(connection);
+            } catch (IOException e) {
+                logger.warn("Send Message Exception: ", e);
+            }
+        }
+
+        @OnWebSocketClose
+        public void onClose(int statusCode, String reason) {
+            logger.warn("Closing a WebSocket due to {}", reason);
+            session = null;
+            connected = false;
+        }
+
+        @OnWebSocketError
+        public void onWebSocketError(Throwable e) {
+            logger.error("Error during websocket communication: {}", e.getMessage(), e);
```

Comment:
The error logging level should be reserved for issues catastrophic enough to threaten the operation of openHab itself. In error in a binding shouldn't qualify for this. Please read the [openHab coding guidelines](https://www.openhab.org/docs/developer/guidelines.html#f-logging) for expected use of logging in bindings.

My general rule for logging levels is:
* trace - used for normal expected execution paths
* debug - used for unexepected execution paths, but well within normal operation
* info - used for notable points in an execution path, like a milestone. (in openhab we try reserve this logging level for the core, so bindings should rarely ever call this.)
* warn - used for notable unexpected execution paths that a regular user (not just a developer) should be notified of. Warnings should be used to indicate that something not-normal occurred and user intervention is required to resolve. Warnings do not indicate a failure to operate merely an abnormal condition of operation that can still be handled by the binding. Failures in binding operation should be indicated by changing the thing status to offline.
* error - used to indicate catastrophic program failure. This should be used to indicate a catastrophic failure in openhab's ability to operate. A failure in a binding would never cause openhab as a whole to fail so a failure in a bindings should never log an error. Instead that failure should be indicated by changing the thing status.

Reply:
Corrected.